src/coreclr/vm/assemblyspec.cpp

```diff
@@ -239,14 +239,11 @@ void AssemblySpec::InitializeSpec(PEAssembly * pFile)
 
     // Set the binding context for the AssemblySpec
     AssemblyBinder* pCurrentBinder = GetBinder();
-    AssemblyBinder* pExpectedBinder = pFile->GetBinder();
     if (pCurrentBinder == NULL)
     {
-        // We should aways have the binding context in the PEAssembly. The only exception to this are the following:
-        //
-        // 1) when we are here during EEStartup and loading CoreLib.
-        // 2) We are dealing with dynamic assemblies
-        _ASSERTE((pExpectedBinder != NULL) || pFile->IsSystem() || pFile->IsDynamic());
+        AssemblyBinder* pExpectedBinder = pFile->GetAssemblyBinder();
+        // We should aways have the binding context in the PEAssembly.
+        _ASSERTE(pExpectedBinder != NULL);
```

Comment:
The original code said that the binder can be null for dynamic assemblies. Did you check that it can't happen anymore?

Reply:
`GetAssemblyBinder` returns `FallbackBinder` for Dynamic assemblies, which is propagated recursively through one that created it. There must be a non-dynamic assembly at least at the root, which would be the System assembly, which now has a binder.

The original comment was correct about dynamic assemblies, but the root cause was still the System not having a binder.

Reply:
Got it - makes sense.